### ISSUE TITLE
Create VisionExtractor service with Florence-2 queries

### ISSUE THREAD

**Issue by mikesvoboda** [mikesvoboda/nemotron-v3-home-security-intelligence#1038]
Create the VisionExtractor service that uses Florence-2 to extract attributes from images.

## Context

VisionExtractor is the core service for querying Florence-2 about image content. It crops detection regions and asks structured questions to extract vehicle attributes, person attributes, and scene context.

### Create backend/services/vision_extractor.py

```python
from dataclasses import dataclass
from PIL import Image
import torch

from backend.services.model_zoo import ModelManager

@dataclass
class VehicleAttributes:
    color: str | None           # "white", "red", "black"
    vehicle_type: str | None    # "sedan", "SUV", "pickup", "van"
    is_commercial: bool
    commercial_text: str | None # "FedEx", "Joe's Plumbing"
    caption: str                # Full description

@dataclass
class PersonAttributes:
    clothing: str | None        # "blue jacket, dark pants"
    carrying: str | None        # "backpack", "package", "nothing"
    is_service_worker: bool
    action: str | None          # "walking", "standing", "crouching"
    caption: str

@dataclass
class SceneAnalysis:
    unusual_objects: list[str]  # ["ladder against fence"]
    tools_detected: list[str]   # ["ladder", "crowbar"]
    abandoned_items: list[str]  # ["package near door"]
    scene_description: str

@dataclass
class EnvironmentContext:
    time_of_day: str            # "day", "dusk", "night"
    artificial_light: bool
    weather: str | None


class VisionExtractor:
    '''Extract visual attributes using Florence-2 vision-language model.'''
    
    VEHICLE_QUERIES = [
        "<CAPTION>",
        "What color is this vehicle?",
        "What type of vehicle is this?",
        "Is this a commercial vehicle?",
        "What company logo or text is visible?",
    ]
    
    PERSON_QUERIES = [
        "<CAPTION>",
        "What is this person wearing?",
        "Is this person carrying anything?",
        "Does this person appear to be a delivery worker or service worker?",
        "What is this person doing?",
    ]
    
    SCENE_QUERIES = [
        "<CAPTION>",
        "Are there any tools visible? (ladder, crowbar, bolt cutters, etc.)",
        "Are there any abandoned bags or packages?",
        "Is there anything unusual or out of place in this scene?",
    ]
    
    ENVIRONMENT_QUERIES = [
        "What time of day does this appear to be based on lighting?",
        "Is there a flashlight or artificial light source visible?",
        "What are the weather conditions?",
    ]
    
    def __init__(self, model_manager: ModelManager):
        self._model_manager = model_manager
    
    async def extract_vehicle_attributes(
        self, 
        image: Image.Image, 
        bbox: tuple[float, float, float, float]
    ) -> VehicleAttributes:
        '''Extract attributes from a vehicle detection.'''
        # Crop vehicle region with padding
        crop = self._crop_with_padding(image, bbox, padding=0.1)
        
        async with self._model_manager.load("florence-2") as (model, processor):
            responses = await self._query_florence(model, processor, crop, self.VEHICLE_QUERIES)
        
        return VehicleAttributes(
            color=self._parse_color(responses[1]),
            vehicle_type=self._parse_vehicle_type(responses[2]),
            is_commercial=self._parse_bool(responses[3]),
            commercial_text=self._parse_commercial_text(responses[4]),
            caption=responses[0],
        )
    
    async def extract_person_attributes(
        self,
        image: Image.Image,
        bbox: tuple[float, float, float, float]
    ) -> PersonAttributes:
        '''Extract attributes from a person detection.'''
        crop = self._crop_with_padding(image, bbox, padding=0.1)
        
        async with self._model_manager.load("florence-2") as (model, processor):
            responses = await self._query_florence(model, processor, crop, self.PERSON_QUERIES)
        
        return PersonAttributes(
            clothing=responses[1],
            carrying=self._parse_carrying(responses[2]),
            is_service_worker=self._parse_bool(responses[3]),
            action=self._parse_action(responses[4]),
            caption=responses[0],
        )
    
    async def analyze_scene(self, image: Image.Image) -> SceneAnalysis:
        '''Analyze full scene for unusual objects.'''
        async with self._model_manager.load("florence-2") as (model, processor):
            responses = await self._query_florence(model, processor, image, self.SCENE_QUERIES)
        
        return SceneAnalysis(
            unusual_objects=self._parse_list(responses[3]),
            tools_detected=self._parse_tools(responses[1]),
            abandoned_items=self._parse_list(responses[2]),
            scene_description=responses[0],
        )
    
    async def analyze_environment(self, image: Image.Image) -> EnvironmentContext:
        '''Analyze environmental context.'''
        async with self._model_manager.load("florence-2") as (model, processor):
            responses = await self._query_florence(model, processor, image, self.ENVIRONMENT_QUERIES)
        
        return EnvironmentContext(
            time_of_day=self._parse_time_of_day(responses[0]),
            artificial_light=self._parse_bool(responses[1]),
            weather=responses[2] if responses[2] else None,
        )
    
    async def _query_florence(
        self,
        model,
        processor,
        image: Image.Image,
        queries: list[str]
    ) -> list[str]:
        '''Run multiple queries against Florence-2.'''
        responses = []
        for query in queries:
            inputs = processor(text=query, images=image, return_tensors="pt").to(model.device)
            
            with torch.no_grad():
                generated_ids = model.generate(
                    **inputs,
                    max_new_tokens=100,
                    num_beams=3,
                )
            
            response = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
            responses.append(response.strip())
        
        return responses
    
    def _crop_with_padding(self, image: Image.Image, bbox: tuple, padding: float) -> Image.Image:
        '''Crop image to bbox with padding.'''
        w, h = image.size
        x1, y1, x2, y2 = bbox
        
        pad_w = (x2 - x1) * padding
        pad_h = (y2 - y1) * padding
        
        x1 = max(0, x1 - pad_w)
        y1 = max(0, y1 - pad_h)
        x2 = min(w, x2 + pad_w)
        y2 = min(h, y2 + pad_h)
        
        return image.crop((int(x1), int(y1), int(x2), int(y2)))
    
    # ... parsing helper methods ...
```

## Parsing Helpers

Implement these parsing methods to normalize Florence-2 responses:

* `_parse_color(text)` - Extract color from response
* `_parse_vehicle_type(text)` - Normalize to sedan/SUV/pickup/van/truck/motorcycle
* `_parse_bool(text)` - Parse yes/no responses
* `_parse_commercial_text(text)` - Extract company names
* `_parse_carrying(text)` - Normalize carrying items
* `_parse_action(text)` - Normalize to walking/standing/crouching/running
* `_parse_time_of_day(text)` - Normalize to day/dusk/night
* `_parse_list(text)` - Split comma-separated items
* `_parse_tools(text)` - Extract tool names

## Tests

Create backend/tests/unit/test_vision_extractor.py:

* Test dataclass creation
* Test \_crop_with_padding geometry
* Test parsing helpers with various inputs
* Test full extraction flow with mock model

## Reference

* Design doc: docs/plans/2026-01-01-vision-extraction-design.md (Florence-2 Extraction Queries section)

---

*Migrated from beads:* `home_security_intelligence-u7bq.3`
*Original created: 2026-01-01*
*Original closed: 2026-01-01*
